### PR TITLE
add containerized mounter option

### PR DESCRIPTION
**Pull request by w9n**
Signed-off-by: Robin Winkelewski <w9ncontact@gmail.com>

**- What I did**
This enables to mount nfs, glusterfs, ceph and cifs volumes by an on runtime deployed root directory on `${CONTAINERIZED_MOUNTER_HOME}/rootfs`, if the variable is set in the kubelet container or the kubelet.sh.conf file and also [${CONTAINERIZED_MOUNTER_HOME}/mounter](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/mounter/mounter.go) is provided. 
**- How I did it**
use the `--experimental-mounter-path` kubelet argument. 
**- How to verify it**
I currently use a hacky daemon-set for this. 
**- Description for the changelog**
add containerized mounter option

linuxkit/linuxkit#2683

**- A picture of a cute animal (not mandatory but encouraged)**
animal

### CODE REVIEW

File: pkg/kubelet/kubelet.sh [hunk 1]
```diff
@@ -63,6 +63,24 @@ echo "kubelet.sh: ${await} has arrived" 2>&1
...
+    mkdir -p "${CONTAINERIZED_MOUNTER_ROOTFS}"/proc
+    mkdir -p "${CONTAINERIZED_MOUNTER_ROOTFS}"/dev
+    mkdir -p "${CONTAINERIZED_MOUNTER_ROOTFS}"/etc
+    touch "${CONTAINERIZED_MOUNTER_ROOTFS}"/etc/resolv.conf
+    mount --bind "${CONTAINERIZED_MOUNTER_HOME}" "${CONTAINERIZED_MOUNTER_HOME}"
+    mount -o remount,exec "${CONTAINERIZED_MOUNTER_HOME}"
+    mount --rbind /var/lib/kubelet "${CONTAINERIZED_MOUNTER_ROOTFS}"/var/lib/kubelet
+    mount --bind -o ro /etc/resolv.conf "${CONTAINERIZED_MOUNTER_ROOTFS}"/etc/resolv.conf
+    mount --bind -o ro /proc "${CONTAINERIZED_MOUNTER_ROOTFS}"/proc
+    mount --bind -o ro /dev "${CONTAINERIZED_MOUNTER_ROOTFS}"/dev
```
Comment (ijc) [inline pkg/kubelet/kubelet.sh]: This looks suspiciously like it really ought to be a service defined in the yml. That might mean enhancing the builder tool and/or init to provide a way to reference one container from another, right now `runtime.BindNS.Mnt` looks like it should be usable for that but it actually isn't. I can't find the issue I raised right now though.

Comment (ijc) [reply pkg/kubelet/kubelet.sh]: Here it is https://github.com/moby/tool/issues/152.

**Comment by w9n** [comment]
I intentionally didnt create a service because that would mean different operating system for different pv clients. In this solutions i can deploy a nfs or ceph filesystem on runtime when it is needed by an app. From what i understand this is the scope of CSI but its not close to an implementation from what i see.

**Comment by ijc** [comment]
Building specific images for specific scenarios would be the LinuxKit model here. 

Although it's not clear to me where the distro comes from (it isn't included here AFAICT) and why it would need to be specific to NFS or CEPH (why not build on which covers both cases if that's what you need).

**Comment by w9n** [comment]
the ceph alpine package is 240mb, debian gluster-client 118mb so i dont want to include them by default. There are also stateless apps that need these utils as client. By deploying them on run time already running nodes can host these apps without rebooting a different image.

I currently have a daemonset with a container where i manually add the root fs with the dependencies. I could contribute it but its basically a rip off from here: https://github.com/kubernetes/kubernetes/tree/master/cluster/gce/gci/mounter

**Comment by ijc** [comment]
You could also pull and launch it dynamically by talking to containerd.

**Comment by justincormack** [comment]
Yeah, I really don't like trying to set up a runnable rootfs by hand. It seems very fragile. There would seem to be better ways of creating a container, eg using `runc` or `containerd` or `docker` to create it (or linuxkit).

**Comment by w9n** [comment]
allright, another nice hack in this solution is if `CONTAINERIZED_MOUNTER_ROOTFS` is a stateful path, then it doesnt have to repull the root dir on reboot, but lets see:

containerd: could be a mounter executable that loads an image defined by the mount parameter from a local yaml and some oci runtime spec in the kubelet container.
- (+) functional easy image (no root dir deployment)
- (+) should be easy to implement
- (-) kubelet needs access to containerd.sock
- (-) images cant be stateful  

runc: same as containerd but operates on root dir.
- (+) no containerd.sock
- (+) root dir can be on stateful path
- (-) gets more complicated in implementation and/or deployment of the root dirs

I dont think docker is an option because its not available in cri-containerd.

I would prefer containerd for simplicity. If you have an elegant example on how to download root dirs from a registry for runc, i would happy to know. I only know [image-tools](https://github.com/opencontainers/image-tools) which depends on skopeo which hasnt implemented Image Manifest Version 2, Schema 2 yet

**Comment by w9n** [comment]
this will be solved by [csi](https://github.com/container-storage-interface/spec).

**Pull request closed without merging**